Clear current message - fixes #1609

rogeralsing opened the pull request:
Fixes #1609

cc @JeffCyr please review that the current message on async `ActorTaskSchedulerMessage` messages looks OK.

rogeralsing commented on src/core/Akka/Actor/ActorCell.DefaultMessages.cs:
```diff
@@ -60,6 +60,7 @@ public void Invoke(Envelope envelope)
                 else
                     ReceiveMessage(message);
+                CurrentMessage = null;
             }
```
From the Scala code

```scala
final def invoke(messageHandle: Envelope): Unit = try {
    currentMessage = messageHandle
    cancelReceiveTimeout() // FIXME: leave this here???
    messageHandle.message match {
      case msg: AutoReceivedMessage ⇒ autoReceiveMessage(messageHandle)
      case msg                      ⇒ receiveMessage(msg)
    }
    ///########## current message is cleared here
    currentMessage = null // reset current message after successful invocation
  } catch handleNonFatalOrInterruptedException { e ⇒
    handleInvokeFailure(Nil, e)
  } finally {
    checkReceiveTimeout // Reschedule receive timeout
  }
```

rogeralsing commented on src/core/Akka/Actor/ActorCell.DefaultMessages.cs:
```diff
@@ -214,6 +215,9 @@ private void HandleActorTaskSchedulerMessage(ActorTaskSchedulerMessage m)
+                //set the current message captured in the async operation
+                //current message was cleared earlier when the async receive handler completed
+                CurrentMessage = m.Message;
                 HandleInvokeFailure(m.Exception);
```
If we are handling a `ActorTaskSchedulerMessage`, we need to re-set the current message captured in the async operation. 
The actorcells current message was cleared when the async receive method completed.

rogeralsing commented on src/core/Akka/Dispatch/ActorTaskScheduler.cs:
```diff
@@ -93,6 +93,7 @@ public static void RunTask(Func<Task> asyncAction)
             //suspend the mailbox
             mailbox.Suspend(MailboxSuspendStatus.AwaitingTask);
+            var currentMessage = context.CurrentMessage;
 
```
Capture the current message from the actor cell

rogeralsing commented on src/core/Akka/Dispatch/ActorTaskScheduler.cs:
```diff
@@ -109,7 +110,7 @@ public static void RunTask(Func<Task> asyncAction)
                                   else
                                   {
-                                      context.Self.Tell(new ActorTaskSchedulerMessage(exception), ActorRefs.NoSender);
+                                      context.Self.Tell(new ActorTaskSchedulerMessage(exception,currentMessage), ActorRefs.NoSender);
```
If the async operation failed, pass back the exception together with the current user message (captured above)

rogeralsing commented on src/core/Akka/Dispatch/SysMsg/ISystemMessage.cs:
```diff
@@ -261,12 +261,14 @@ public ActorTaskSchedulerMessage(ActorTaskScheduler scheduler, Task task)
         {
             Exception = exception;
+            Message = message;
         }
```
`ActorTaskSchedulerMessage`  now also carries the user message that caused the exception.

rogeralsing commented:
Some notes.
The rationale for the implementation here is that we should stick to how Scala does this, and all the magic of async operations is dealt with externally.
We set and clear the current message on entry and exit of the `ActorCell.Invoke`.
When doing this, we also need to make sure that async operations have access to the current message in their callbacks, this is solved by capturing the current user message when an async operation starts, and then pass that message back to the actor cell using the `ActorTaskSchedulerMessage` message.

JeffCyr commented:
ActorCell.CurrentMessage won't be set after the first await in the async receive. It will be restored properly before calling HandleInvokeFailure in case of an exception, but it won't be set for other usage:

AbstractStash.Stash
DefaultResizer.Pressure
PersistentView.ReplayStarted

Stash won't work after the first await, not sure how it impacts the other usages.

This could be fixed by creating an ActorTaskScheduler instance per async receive, the current message would be kept in the instance so it can be restored on each ActorTaskSchedulerMessage handling.

An instance per message is not that bad for async receive, the throughput of async receive should not be high anyway. However this [issue](https://github.com/dotnet/coreclr/issues/2189) may impact the performance of this approach.

An alternative would be to use a custom ActorSynchronizationContext instead of an ActorTaskScheduler.

rogeralsing commented:
>ActorCell.CurrentMessage won't be set after the first await in the async receive. 

I have updated the code to restore the current message on all ActorTaskSchedulerMessage messages.

JeffCyr commented:
Looks good!

JeffCyr commented on src/core/Akka/Dispatch/ActorTaskScheduler.cs:
```diff
@@ -57,8 +57,8 @@ protected override void QueueTask(Task task)
-        {
-            _actorCell.Self.Tell(new ActorTaskSchedulerMessage(this, task), ActorRefs.NoSender);
+        {            
+            _actorCell.Self.Tell(new ActorTaskSchedulerMessage(this, task, _actorCell.CurrentMessage), ActorRefs.NoSender);
```
Hmm wait, I think this approach doesn't work if the task is scheduled from outside of the actor's context.

rogeralsing replied:
AFAIK, no one could ever use this outside of the actor context as each scheduler is bound to an individual actor and only used from within a context.
Are there any cases I am missing?

JeffCyr replied:
```c#
Receive(async m =>
{
    // ...
    await Task.Delay(1000);

});
```

JeffCyr commented:
@rogeralsing Just saw your edit, that's right ActorTaskScheduler could capture CurrentMessage in a field as long as it is cleared when the async receive completes to not recreate the original issue.

rogeralsing commented on src/core/Akka/Dispatch/ActorTaskScheduler.cs:
```diff
@@ -109,9 +112,10 @@ public static void RunTask(Func<Task> asyncAction)
                                   }
-
+                                  //clear the current message field of the scheduler
+                                  actorScheduler.CurrentMessage = null;
```
Clear the current message of the scheduler

rogeralsing commented on src/core/Akka/Dispatch/SysMsg/ISystemMessage.cs:
```diff
@@ -251,22 +251,26 @@ internal sealed class ActorTaskSchedulerMessage : ISystemMessage
         }
 
         public Exception Exception { get; private set; }
+        public object Message { get;private set; }
```
Placeholder for the current user message inside the scheduler

rogeralsing commented on src/core/Akka.Tests/Dispatch/AsyncAwaitSpec.cs:
```diff
@@ -57,10 +57,14 @@ public SuspendActor()
+                var current = cell.CurrentMessage;
+                //ensure we have the correct current message here
+                Assert.Same(m_, current);
                 //we expect that state should not have changed due to an incoming message
```
Assert that we in fact do have the correct current message after awaiting a Task.Delay

rogeralsing commented:
cc @JeffCyr , I've updated the scheduler to keep a field for the current message.
I've also updated the tests to ensure that we do have the correct current message after awaiting a Task.Delay.

JeffCyr commented:
Looks good! :thumbsup:

Aaronontheweb commented:
Should we add a spec to ensure that the current message is cleared?

rogeralsing commented on src/core/Akka.Tests/Actor/ActorCellSpec.cs:
```diff
@@ -0,0 +1,40 @@
+            var actor = Sys.ActorOf(Props.Create(() => new DummyActor(autoResetEvent)));
+            actor.Tell("hello");
+            //ensure the message was received
+            autoResetEvent.WaitOne();
```
wait until the actor have received our message

rogeralsing commented on src/core/Akka.Tests/Actor/ActorCellSpec.cs:
```diff
@@ -0,0 +1,40 @@
+            //ensure the message was received
+            autoResetEvent.WaitOne();
+            var refCell = actor as ActorRefWithCell;
+            var cell = refCell.Underlying as ActorCell;
```
get the actorcell from the actorref

rogeralsing commented on src/core/Akka.Tests/Actor/ActorCellSpec.cs:
```diff
@@ -0,0 +1,40 @@
+            //wait while current message is not null (that is, receive is not yet completed/exited)
+            SpinWait.SpinUntil(() => cell.CurrentMessage != null, TimeSpan.FromSeconds(2));
+
+            cell.CurrentMessage.ShouldBe(null);
```
wait for the current message to become null, or timeout
then assert that current message is null

rogeralsing commented:
@Aaronontheweb I've added a spec to ensure that the current message is cleared.

rogeralsing commented on src/core/Akka.Tests/Actor/ActorCellSpec.cs:
```diff
@@ -0,0 +1,68 @@
+            //wait while current message is not null (that is, receive is not yet completed/exited)
+            SpinWait.SpinUntil(() => cell.CurrentMessage == null, TimeSpan.FromSeconds(2));
+
+            cell.CurrentMessage.ShouldBe(null);
```
wait for current message to become null, or timeout
Then assert that current message is null

Aaronontheweb replied:
You can use `AwaitAssert` for this in the future, but this change is fine

rogeralsing commented on src/core/Akka.Tests/Actor/ActorCellSpec.cs:
```diff
@@ -0,0 +1,68 @@
+            //wait while current message is not null (that is, receive is not yet completed/exited)
+            SpinWait.SpinUntil(() => cell.CurrentMessage == null, TimeSpan.FromSeconds(2));
+
+            cell.CurrentMessage.ShouldBe(null);
```
Same test as above but for async await actor

Aaronontheweb commented:
Looks like #1608 dirtied the working directory and failed this spec, which is bad. This is an issue with how I've configured TeamCity, not anything to do with our code. I'll get to work on fixing this.

Aaronontheweb commented:
All of the tests pass - will review this when I get back from a meeting

Aaronontheweb commented:
Looks fine to me.
